src/details/ArborX_Callbacks.hpp

```diff
@@ -44,6 +46,25 @@ struct DefaultCallback
   }
 };
 
+#ifdef ARBORX_ENABLE_MPI
+struct ConstrainedNearestCallbackTag
+{};
+
+struct DefaultCallbackWithRank
+{
+  using tag = ConstrainedNearestCallbackTag;
+
```

Comment:
@dalg24 I know you don't like this approach (similar to the discussion we had in #675). What alternative do you propose?

Reply:
You define a decorator that wraps the callback and you check whether the callback is decorated.

Something like
```C++
template <class Callback>
struct DecoratedCallback {
  Callback callback_;
  template <class... Args>
  void operator()(Args&&... args) const { callback_((Args&&)args...); }
};

template <class Callback>
struct is_decorated_callback : std::false_type {};

template <class Callback>
struct is_decorated_callback<DecorateCallback<Callback>> : std::true_type {};
```
Try to come up with more descriptive names.
I am not convinced this belongs to the general callback header, you might want to move it to the distributed headers.

Reply:
Opted for the decorator instead of updating `operator()`.